### PR TITLE
implemented LengthFieldBaseFrameDecoder

### PR DESCRIPTION
**Pull request by Aaronontheweb**
Implemented the following classes:
- LengthFieldBaseFrameDecoder
- TooLongFrameException
- CorruptedFrameException

Added the following projects:
- DotNetty.Codecs.Tests

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
Need to implement `EmbeddedChannel` in order to write specs for the codecs. Started working on that.

File: src/DotNetty.Codecs/LengthFieldBasedFrameDecoder.cs [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+  * | HDR1 | Length | HDR2 | Actual Content |----->| HDR2 | Actual Content |
+  * | 0xCA | 0x0010 | 0xFE | "HELLO, WORLD" |      | 0xFE | "HELLO, WORLD" |
+  * +------+--------+------+----------------+      +------+----------------+
+  * </pre>
+  * @see LengthFieldPrepender
+  */
+    public class LengthFieldBasedFrameDecoder : ByteToMessageDecoder
+    {
+        readonly ByteOrder _byteOrder;
+        readonly int _maxFrameLength;
```
Comment (nayato) [inline src/DotNetty.Codecs/LengthFieldBasedFrameDecoder.cs]: please, remove leading underscore from field names for consistency.

File: src/DotNetty.Codecs/LengthFieldBasedFrameDecoder.cs [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+                case 8:
+                    frameLength = buffer.GetLong(offset);
+                    break;
+                default:
+                    throw new DecoderException("unsupported lengthFieldLength: " + this._lengthFieldLength + " (expected: 1, 2, 3, 4, or 8)");
+            }
+            return frameLength;
+        }
+
+        protected IByteBuffer ExtractFrame(IChannelHandlerContext context, IByteBuffer buffer, int index, int length)
```
Comment (nayato) [inline src/DotNetty.Codecs/LengthFieldBasedFrameDecoder.cs]: [nit] will become virtual to support overrides in MarshallingDecoder and ObjectDecoder

File: src/DotNetty.Codecs/LengthFieldBasedFrameDecoder.cs [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+        {
+            var buff = buffer.Slice(index, length);
+            buff.Retain();
+            return buff;
+        }
+
+        private void FailIfNecessary(bool firstDetectionOfTooLongFrame)
```
Comment (nayato) [inline src/DotNetty.Codecs/LengthFieldBasedFrameDecoder.cs]: [style] can leave private out as it's the for class members.

**Comment by nayato** [comment]
small things, otherwise looks good!

**Comment by Aaronontheweb** [comment]
Done - updated + squashed commits.

**Pull request merged**